perf_bench_test.go

```diff
@@ -24,127 +24,240 @@ package main
+			go serveVM(t, vmIDString, imageName, &vmGroup, isSyncOffload, &avgExecTime, &realRPS)
 
 			vmID = (vmID + 1) % *vmNum
-		case <-done:
-			ticker.Stop()
+			if vmID == 0 {
+				perfStat.SetWarmTime()
+			}
 		}
 	}
 
-	done <- true
+	ticker.Stop()
 	vmGroup.Wait()
+	perfStat.SetTearDownTime()
+
+	// Shutdown VMs
+	message, err := funcPool.RemoveInstance("plr_fnc", images[0], isSyncOffload)
+	require.NoError(t, err, "Function returned error, "+message)
+	for i := 0; i < *vmNum; i++ {
+		vmIDString := strconv.Itoa(i)
+		message, err := funcPool.RemoveInstance(vmIDString, images[i%len(images)], isSyncOffload)
+		require.NoError(t, err, "Function returned error, "+message)
+	}
+
+	// Collect results
+	serveMetrics := make(map[string]float64)
+	serveMetrics[averageExecutionTime] = float64(avgExecTime) / float64(totalRequests)
+	serveMetrics[realRequestsPerSecond] = float64(realRPS)
+
+	if isRunPerf {
+		result, err := perfStat.GetResult()
+		require.NoError(t, err, "Stop perf stat returned error: %v", err)
+		for eventName, value := range result {
+			log.Debugf("%s: %f\n", eventName, value)
+			serveMetrics[eventName] = value
+		}
+	}
+	log.Debugf("average-execution-time: %f\n", serveMetrics[averageExecutionTime])
+	log.Debugf("real-requests-per-second: %f\n", serveMetrics[realRequestsPerSecond])
+
+	writeResultToCSV(t, "benchRPS.csv", serveMetrics)
 }
 
-///////////////////////////////////////////////////////////////////////////////
-////////////////////////// Auxialiary functions below /////////////////////////
-///////////////////////////////////////////////////////////////////////////////
-func serveVM(t *testing.T, start time.Time, vmIDString, imageName string, vmGroup *sync.WaitGroup, isSyncOffload bool) {
+func serveVM(t *testing.T, vmIDString, imageName string, vmGroup *sync.WaitGroup, isSyncOffload bool, avgExecTime, realRPS *int64) {
 	defer vmGroup.Done()
 
+	tStart := time.Now()
 	resp, _, err := funcPool.Serve(context.Background(), vmIDString, imageName, "replay")
-	require.NoError(t, err, "Function returned error")
-	require.Equal(t, resp.Payload, "Hello, replay_response!")
-
-	if *isColdStart {
-		require.Equal(t, resp.IsColdStart, true)
-		message, err := funcPool.RemoveInstance(vmIDString, imageName, isSyncOffload)
-		require.NoError(t, err, "Function returned error, "+message)
+	require.Equal(t, resp.IsColdStart, false)
+	if err != nil {
+		log.Warn("Function returned error")
 	} else {
-		require.Equal(t, resp.IsColdStart, false)
+		atomic.AddInt64(realRPS, 1)
+	}
+	if resp.Payload != "Hello, replay_response!" {
+		log.Warnf("Function returned invalid: %s", resp.Payload)
+	}
+
+	execTime := time.Since(tStart).Milliseconds()
+	atomic.AddInt64(avgExecTime, execTime)
+	log.Debugf("VM %s: returned in %d milliseconds", vmIDString, execTime)
+}
+
+// Returns a list of image names
+func getImages(t *testing.T) []string {
+	var (
+		images = map[string]string{
+			"helloworld":   "ustiugov/helloworld:var_workload",
```

Comment:
use `vhiveease/helloworld:var_workload` instead of `ustiugov/helloworld:var_workload` for all functions

Reply:
Done